pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3434 @@
+        node, which is used in tree visitor scripts.  In the simplest
+        case, this simply returns::
+
+            self.__class__(args))
+
+        But in general this constructs a copy of the current
+        expression object using local data as well as arguments
+        that represent the child nodes.   Thus, this method can be viewed
+        as a node-level clone method.
+
+        Args:
+            args (list): A list of child nodes for the new expression
+                object
+            memo (dict): A dictionary that maps object ids to clone
+                objects generated earlier during a cloning process.
+                This argument is needed to clone objects that are
+                owned by a model, and it can be safely ignored for
+                most expression classes.
+
+        Returns:
+            A new expression object with the same type as the current
+            class.
+        """
+        return self.__class__(args)
+
+    def is_constant(self):
+        """Return True if this expression is an atomic constant
+
+        This method contrasts with the is_fixed() method.  This method
+        returns True if the expression is an atomic constant, that is it
+        is composed exclusively of constants and immutable parameters.
+        NumericValue objects returning is_constant() == True may be
+        simplified to their numeric value at any point without warning.
+
+        Note:  This defaults to False, but gets redefined in sub-classes.
+        """
+        return False
+
+    def is_fixed(self):
+        """
+        Return :const:`True` if this expression contains no free variables.
+
+        Returns:
+            A boolean.
+        """
+        return _expression_is_fixed(self)
+
+    def _is_fixed(self, values):
+        """
+        Compute whether this expression is fixed given
+        the fixed values of its children.
+
+        This method is called by the :class:`_IsFixedVisitor
+        <pyomo.core.expr.current._IsFixedVisitor>` class.  It can
+        be over-written by expression classes to customize this
+        logic.
+
+        Args:
+            values (list): A list of boolean values that indicate whether
+                the children of this expression are fixed
+
+        Returns:
+            A boolean that is :const:`True` if the fixed values of the
+            children are all :const:`True`.
+        """
+        return all(values)
+
+    def is_potentially_variable(self):
+        """
+        Return :const:`True` if this expression contains variables.
+
+        This method returns :const:`True` if there are any variables
+        within this expression.  This method returns :const:`True`
+        even if there the variables in the expression are currently
+        fixed.
+
+        Returns:
+            A boolean.  Defaults to :const:`True` for expressions.
+        """
+        return True
+
+    def is_named_expression_type(self):
```

Comment:
This seems like a *very* specific method.  Why do we need it?  In particular, since Expressions are components, could/should we not use ctype?

Reply:
Named expressions get treated differently during cloning.  This method is needed to support duck typing for named expressions.

Reply:
If we get rid of the expression-specific clone operation (and rely entirely on deepcopy), then named `Expression` objects will be handled correctly automatically.  I think that would allow us to remove this method.